In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
import os

In [2]:
#reading data

df1 = pd.read_csv("C:\\Users\\gagan\\Desktop\\Data Set\\phone_user_review_file_1.csv",encoding='iso-8859-1')
df2 = pd.read_csv("C:\\Users\\gagan\\Desktop\\Data Set\\phone_user_review_file_2.csv",encoding='iso-8859-1')
df3 = pd.read_csv("C:\\Users\\gagan\\Desktop\\Data Set\\phone_user_review_file_3.csv",encoding='iso-8859-1')
df4 = pd.read_csv("C:\\Users\\gagan\\Desktop\\Data Set\\phone_user_review_file_4.csv",encoding='iso-8859-1')
df5 = pd.read_csv("C:\\Users\\gagan\\Desktop\\Data Set\\phone_user_review_file_5.csv",encoding='iso-8859-1')
df6 = pd.read_csv("C:\\Users\\gagan\\Desktop\\Data Set\\phone_user_review_file_6.csv",encoding='iso-8859-1')

In [3]:
#comparing shape of all data
s1=df1.shape
s2=df2.shape
s3=df3.shape
s4=df4.shape
s5=df5.shape
s6=df6.shape
print(s1,s2,s3,s4,s5,s6)

(374910, 11) (114925, 11) (312961, 11) (98284, 11) (350216, 11) (163837, 11)


all the data set have same column

In [4]:
#merging all data into one

df = pd.concat([df1,df2,df3,df4,df5,df6],ignore_index=True)
df.head(10)

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   
5  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
6  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
7  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
8  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
9  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   
5  verizonwireless.com   10.0       10.0   
6  verizonwireless.com    4.0       10.0   
7  verizonwireless.com    6.0       10.0   
8  verizonwireless.com    6.0       10.0   
9  verizonwireless.com    8.0       10.0   

                                             extract           author  \
0  As a diehard Samsung fan who has had every Sam...       CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...        james0923   
2  Adequate feel. Nice heft. Processor's still sl...         R. Craig   
3  Never disappointed. One of the reasons I've be...      Buster2020    
4  I've now found that i'm in a group of people t...       S Ate Mine   
5  I am the type of person who never would comple...            BDB76   
6  The way this Samsung S8 phone operates is more...         KLC30306   
7  I bought this phone very excited to use it. I ...             Mnhy   
8  It is an extremely advanced and truly a Smart ...          Mr Alan   
9  Great phone with a PHENOMENAL Camera, not all ...  Raven Wolf Song   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8  
5                                  Samsung Galaxy S8  
6                                  Samsung Galaxy S8  
7                                  Samsung Galaxy S8  
8                                  Samsung Galaxy S8  
9                                  Samsung Galaxy S8

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1415133 entries, 0 to 1415132
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1415133 non-null  object 
 1   date       1415133 non-null  object 
 2   lang       1415133 non-null  object 
 3   country    1415133 non-null  object 
 4   source     1415133 non-null  object 
 5   domain     1415133 non-null  object 
 6   score      1351644 non-null  float64
 7   score_max  1351644 non-null  float64
 8   extract    1395772 non-null  object 
 9   author     1351931 non-null  object 
 10  product    1415132 non-null  object 
dtypes: float64(2), object(9)
memory usage: 118.8+ MB


In [6]:
df.describe().transpose()

count      mean       std   min   25%   50%   75%   max
score      1351644.0   8.00706  2.616121   0.2   7.2   9.2  10.0  10.0
score_max  1351644.0  10.00000  0.000000  10.0  10.0  10.0  10.0  10.0

score_max have std deviation of 0 which means it has only one unique value.

In [7]:
df.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
dtype: int64

null values are :

score        63489 (4.49%)
score_max    63489 (4.49%)
extract      19361 (1.37%)
author       63202 (4.47%)

In [8]:
#impute missing values with median
df['score'] = df['score'].fillna(df['score'].median())

In [9]:
df.nunique()

phone_url       5556
date            7728
lang              22
country           42
source           331
domain           384
score             86
score_max          1
extract      1321353
author        801103
product        61313
dtype: int64

score_max has only one value

In [10]:
df['score']= round(df['score'])
df.head(10)

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   
5  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
6  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
7  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
8  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
9  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.0       10.0   
4  verizonwireless.com    4.0       10.0   
5  verizonwireless.com   10.0       10.0   
6  verizonwireless.com    4.0       10.0   
7  verizonwireless.com    6.0       10.0   
8  verizonwireless.com    6.0       10.0   
9  verizonwireless.com    8.0       10.0   

                                             extract           author  \
0  As a diehard Samsung fan who has had every Sam...       CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...        james0923   
2  Adequate feel. Nice heft. Processor's still sl...         R. Craig   
3  Never disappointed. One of the reasons I've be...      Buster2020    
4  I've now found that i'm in a group of people t...       S Ate Mine   
5  I am the type of person who never would comple...            BDB76   
6  The way this Samsung S8 phone operates is more...         KLC30306   
7  I bought this phone very excited to use it. I ...             Mnhy   
8  It is an extremely advanced and truly a Smart ...          Mr Alan   
9  Great phone with a PHENOMENAL Camera, not all ...  Raven Wolf Song   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8  
5                                  Samsung Galaxy S8  
6                                  Samsung Galaxy S8  
7                                  Samsung Galaxy S8  
8                                  Samsung Galaxy S8  
9                                  Samsung Galaxy S8

In [11]:
# checking duplicate values
df.duplicated().sum()

6436

there are 6436 duplicate values in data

In [12]:
#dropping duplicate values
df = df.drop_duplicates()
df.shape

(1408697, 11)

after dropping duplicate values we have 1408697 columns and 11 rows

In [13]:
#dropping irrelevant features

df=df.drop(['phone_url', 'date', 'lang', 'country', 'source', 'domain', 'score_max', 'extract'], axis=1)
df.head()

score       author                                            product
0   10.0   CarolAnn35                                  Samsung Galaxy S8
1   10.0    james0923                                  Samsung Galaxy S8
2    6.0     R. Craig  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...
3    9.0  Buster2020                       Samsung Galaxy S8 64GB (AT&T)
4    4.0   S Ate Mine                                  Samsung Galaxy S8

In [14]:
# 1 million data samples
df = df.sample(n=1000000, random_state=612)
df.shape

(1000000, 3)

In [15]:
#Identify the most rated products

df['product'].value_counts().head()

Lenovo Vibe K4 Note (White,16GB)     3709
Lenovo Vibe K4 Note (Black, 16GB)    3083
OnePlus 3 (Graphite, 64 GB)          2890
OnePlus 3 (Soft Gold, 64 GB)         2522
Samsung Galaxy Express I8730         1898
Name: product, dtype: int64

In [16]:
#Identify the users with most number of reviews.
df['author'].value_counts().head()

Amazon Customer    54542
Cliente Amazon     13661
e-bit               5959
Client d'Amazon     5495
Amazon Kunde        3283
Name: author, dtype: int64

In [17]:
#data with products having more than 50 ratings and users who have given more than 50 ratings

# list of authors with > 50 ratings
auth = df['author'].value_counts()
auth = auth[auth>50].index.tolist() 
len(auth)

650

In [18]:
# list of products with > 50 ratings
product = df['product'].value_counts()
product = product[product>50].index.tolist()
len(product)

4250

In [19]:
#The result will only be true at a location if all the labels match.

df_fifty = df[(df['author'].isin(auth)) & (df['product'].isin(product))]
df_fifty.shape

(101504, 3)

## A popularity based model that will recommend top 5 mobile phones

In [20]:
df.groupby('product')['score'].mean().sort_values(ascending=False).head()

product
'Sony Xperia X (F5122) – White – Dual Sim (Google Android 6.0.1, 5 Display, 2 x CORTEX A72 1.8 GHz + 4 x cortex-a53...                                                         10.0
Elegiant THL 5000 Smartphone 5.0 "Android4.4 MTK6592T, Octa-core, 2.0GHz Dual SIM card dual standby ROM: 16 GB / RAM: 2GB Ultraphone WCDMA 850 / 2100MHz                           10.0
Samsung Galaxy Fame S6810 Unlocked GSM Android Smartphone - Blue                                                                                                                   10.0
Samsung Galaxy Fame Lite GT-S6790 - Smartphone (88.9 mm (3.5 "), 480 x 320 Pixeles, TFT, 0.85 GHz, 4 GB, microSD (TransFlash)) (importado)                                         10.0
Samsung Galaxy Fame LITE - Black - Smartphone con display TFT da 8,9 cm (3,5"), 850 MHz, 512 MB RAM, fotocamera da 3.2 megapixel, 4 GB di memoria interna, USB 2.0, Android 4.1    10.0
Name: score, dtype: float64

In [21]:
df.groupby('product')['score'].count().sort_values(ascending=False).head() 

product
Lenovo Vibe K4 Note (White,16GB)     3709
Lenovo Vibe K4 Note (Black, 16GB)    3083
OnePlus 3 (Graphite, 64 GB)          2890
OnePlus 3 (Soft Gold, 64 GB)         2522
Samsung Galaxy Express I8730         1898
Name: score, dtype: int64

In [22]:
rating_mean_count = pd.DataFrame(df.groupby('product')['score'].mean())
rating_mean_count['rating_counts'] = df.groupby('product')['score'].count()
rating_mean_count = rating_mean_count.sort_values(by=['score','rating_counts'], ascending=[False,False])

In [23]:
rating_mean_count.head() 

score  rating_counts
product                                                                 
Motorola Smartphone Motorola Moto X Desbloquead...   10.0            140
Motorola Smartphone Motorola Moto G Dual Chip D...   10.0            128
Samsung Galaxy Note5                                 10.0            127
Samsung Smartphone Dual Chip Samsung Galaxy SII...   10.0            122
Nokia Smartphone Nokia Lumia 520 Desbloqueado O...   10.0            116

In [24]:
rating_mean_count = pd.DataFrame(df_fifty.groupby('product')['score'].mean())
rating_mean_count['rating_counts'] = df_fifty.groupby('product')['score'].count()
rating_mean_count = rating_mean_count.sort_values(by=['score','rating_counts'], ascending=[False,False])

In [25]:
rating_mean_count.head() 

score  rating_counts
product                                                                 
Samsung Galaxy Note5                                 10.0             50
Sim Free Samsung Galaxy S7 Edge Mobile Phone - ...   10.0             16
Samsung Galaxy S III, Purple 16GB (Sprint)           10.0             10
BlackBerry Passport Smartphone (11,4 cm (4,5 Zo...   10.0              9
Samsung Galaxy S3                                    10.0              9

## collaborative filtering model using KNNWithMeans item_based

In [31]:
from surprise import KNNWithMeans
from surprise import Dataset
from surprise import accuracy
from surprise.model_selection import train_test_split
from surprise import Reader

df_co = Dataset.load_from_df(df_fifty[['author','product','score']], Reader(rating_scale=(1, 10)))
trainset, testset = train_test_split(df_co, test_size=.20,random_state=612)

In [32]:
algo = KNNWithMeans(k=50, sim_options={'name': 'pearson_baseline', 'user_based': False})
algo.fit(trainset)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


In [34]:
test_pred = algo.test(testset)

In [35]:
# get RMSE
print("Item-based Model : Test Set")
accuracy.rmse(test_pred, verbose=True)

Item-based Model : Test Set
RMSE: 2.6827


2.682723823684798

## collaborative filtering model using KNNWithMeans user_based

In [38]:
algo1 = KNNWithMeans(k=50, sim_options={'name': 'pearson_baseline', 'user_based': True})
algo1.fit(trainset)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


In [39]:
test_pred1= algo1.test(testset)

In [40]:
# get RMSE
print("Item-based Model : Test Set")
accuracy.rmse(test_pred1, verbose=True)

Item-based Model : Test Set
RMSE: 2.7370


2.7369692948859905

## a collaborative filtering model using SVD

In [41]:
from surprise import SVD
from surprise.model_selection import cross_validate

svd = SVD(random_state=612)
svd.fit(trainset)
svd_pred = svd.test(testset)

In [42]:
accuracy.rmse(svd_pred)

RMSE: 2.6525


2.6525161672981636

## Average ratings for test users 

In [43]:
svd_pred_df=pd.DataFrame(svd_pred, columns=['uid', 'iid', 'rui', 'est', 'details'])
print('average prediction for test users: ',svd_pred_df['est'].mean())
print('average rating  by test users: ',svd_pred_df['rui'].mean())
print('average prediction error for test users: ',(svd_pred_df['rui']-svd_pred_df['est']).abs().mean())

average prediction for test users:  7.88899483235836
average rating  by test users:  7.84907147431161
average prediction error for test users:  2.0412485395048763


In [46]:
knn_i_df=pd.DataFrame(test_pred, columns=['uid', 'iid', 'rui', 'est', 'details'])
print('average prediction for test users: ',knn_i_df['est'].mean())
print('average rating  by test users: ',knn_i_df['rui'].mean())
print('average prediction error for test users: ',(knn_i_df['rui']-knn_i_df['est']).abs().mean())

average prediction for test users:  7.855419528433595
average rating  by test users:  7.84907147431161
average prediction error for test users:  2.050831029113551


In [47]:
knn_u_pred_df=pd.DataFrame(test_pred1, columns=['uid', 'iid', 'rui', 'est', 'details'])
print('average prediction for test users: ',knn_u_pred_df['est'].mean())
print('average rating  by test users: ',knn_u_pred_df['rui'].mean())
print('average prediction error for test users: ',(knn_u_pred_df['rui']-knn_u_pred_df['est']).abs().mean())

average prediction for test users:  7.8731241443030004
average rating  by test users:  7.84907147431161
average prediction error for test users:  2.0614511972517016


## Report your findings and inferences

1. Many users(author) have given the rating of 10.
2. All three models have alost similar RMSE vlaue

## Try and recommend top 5 products for test users

In [54]:
from collections import defaultdict

top_n = defaultdict(list)
for uid, iid, true_r, est, _ in test_pred:
    top_n[uid].append((iid, est))

for uid, user_ratings in top_n.items():
    user_ratings.sort(key=lambda x: x[1], reverse=True)
    top_n[uid] = user_ratings[:5]


In [55]:
top_n

defaultdict(list,
            {'Pablo': [('Apple iPhone 5 32GB', 10),
              ('Lenovo Motorola Moto G 8 GB - Smartphone libre Android (pantalla 4.5", cámara 5 Mp, 8 GB, Quad-Core 1.2 GHz), negro',
               9.593808327581218),
              ('Huawei P8 Lite - Smartphone de 5" (cámara 13 MP, 16 GB, HiSilicon Kirin 620 Octa Core 1.2 GHz, 2 GB RAM, Android L), color blanco',
               9.388783137122521),
              ('Motorola V3', 9.333333333333334),
              ('Nokia E71 Smartphone (UMTS, WLAN, A-GPS, Bluetooth, 3,2 MP, Ovi Karten, QWERTZ-Tastatur) white steel',
               9.004203177855917)],
             'Ð\x9dадежда': [('Sony Xperia Z1 (Ñ\x84иолеÑ\x82овÑ\x8bй)',
               9.486210410131655),
              ('Sony Xperia Z1 (Ñ\x84иолеÑ\x82овÑ\x8bй)',
               9.486210410131655),
              ('Sony Xperia Z1 (белÑ\x8bй)', 9.422941859703359),
              ('LG X power (индиго)', 9.104979798832124),
              ('S

## Cross validation

In [58]:
svd_cv = cross_validate(svd,df_co, measures=['RMSE'], cv=5, verbose=True)
svd_cv

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.6556  2.6529  2.6580  2.6677  2.6519  2.6572  0.0056  
Fit time          0.80    0.87    0.86    0.86    0.91    0.86    0.04    
Test time         0.25    0.11    0.11    0.10    0.09    0.13    0.06    


{'test_rmse': array([2.65560887, 2.65294165, 2.65800722, 2.66766587, 2.65193738]),
 'fit_time': (0.7968704700469971,
  0.8726761341094971,
  0.8565382957458496,
  0.8572182655334473,
  0.9063451290130615),
 'test_time': (0.24723482131958008,
  0.10957098007202148,
  0.10944628715515137,
  0.10048031806945801,
  0.09365344047546387)}

In [59]:
knn_i_cv = cross_validate(algo,df_co, measures=['RMSE'], cv=5, verbose=False)
knn_i_cv

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


{'test_rmse': array([2.73143677, 2.70619931, 2.75517191, 2.73692634, 2.7568758 ]),
 'fit_time': (1.1142668724060059,
  1.0138270854949951,
  1.0413427352905273,
  1.158614158630371,
  1.1802899837493896),
 'test_time': (2.9693472385406494,
  2.5887725353240967,
  2.6216342449188232,
  3.3513524532318115,
  3.1463024616241455)}

In [60]:
knn_u_cv = cross_validate(algo1,df_co, measures=['RMSE'], cv=5, verbose=False)
knn_u_cv

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


{'test_rmse': array([2.74013689, 2.71954806, 2.74952444, 2.75951605, 2.72892857]),
 'fit_time': (1.2060794830322266,
  1.2072632312774658,
  1.050520896911621,
  1.0998892784118652,
  1.1047677993774414),
 'test_time': (3.5149850845336914,
  2.3541929721832275,
  2.2389533519744873,
  2.2875137329101562,
  2.354247808456421)}

### In what business scenario you should use popularity based Recommendation Systems ?

1. In the condition of cold start.
2. We can use popularity based Recommendation Systems for trends.

### In what business scenario you should use CF based Recommendation Systems

1. By using CF based Recommendation Systems, personalised recommendation can be given to the user.
2. Collaborative filtering can help recommenders to not overspecialize in a user’s profile.

### What other possible methods can you think of which can further improve the recommendation for different users ? 

By leveraging tools like Event Stream Processing and Intelligent Decisioning, and operationalize the entire recommendation process.